naga/src/back/hlsl/writer.rs

```diff
@@ -1213,6 +1227,53 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
             self.write_workgroup_variables_initialization(func_ctx, module)?;
         }
 
+        if let back::FunctionType::EntryPoint(ep_index) = func_ctx.ty {
+            let ep = &module.entry_points[ep_index as usize];
+            for (index, arg) in func.arguments.iter().enumerate() {
+                if let Some(crate::Binding::BuiltIn(builtin)) = arg.binding {
+                    if matches!(
+                        builtin,
+                        crate::BuiltIn::SubgroupSize
+                            | crate::BuiltIn::SubgroupInvocationId
+                            | crate::BuiltIn::NumSubgroups
+                            | crate::BuiltIn::SubgroupId
+                    ) {
+                        let level = back::Level(1);
+                        write!(self.out, "{level}const ")?;
+
+                        self.write_type(module, arg.ty)?;
+
+                        let argument_name =
+                            &self.names[&NameKey::EntryPointArgument(ep_index, index as u32)];
+                        write!(self.out, " {argument_name} = ")?;
+
+                        match builtin {
+                            crate::BuiltIn::SubgroupSize => {
+                                writeln!(self.out, "WaveGetLaneCount();")?
+                            }
+                            crate::BuiltIn::SubgroupInvocationId => {
+                                writeln!(self.out, "WaveGetLaneIndex();")?
+                            }
+                            crate::BuiltIn::NumSubgroups => writeln!(
+                                self.out,
+                                "({}u + WaveGetLaneCount() - 1u) / WaveGetLaneCount();",
+                                ep.workgroup_size[0] * ep.workgroup_size[1] * ep.workgroup_size[2]
+                            )?,
+                            crate::BuiltIn::SubgroupId => {
+                                writeln!(
+                                    self.out,
+                                    "(__local_invocation_id.z * {}u + __local_invocation_id.y * {}u + __local_invocation_id.x) / WaveGetLaneCount();",
+                                    ep.workgroup_size[0] * ep.workgroup_size[1],
+                                    ep.workgroup_size[1],
+                                )?;
+                            }
```

Comment:
The SPIR-V spec (which can be targeted by both GLSL and HLSL) says it is guaranteed that:
`LocalInvocationIndex = LocalInvocationID.z * WorkGroupSize.x * WorkGroupSize.y + LocalInvocationID.y * WorkGroupSize.x + LocalInvocationID.x;`

And if `WorkGroupSize.x % SubgroupSize == 0` then it is also guaranteed that:
`LocalInvocationIndex = SubgroupLocalInvocationId + SubgroupId * SubgroupSize`

Reply:
I think the problem is that it's not guaranteed that the value returned by `WaveGetLaneCount()` is static across the whole workgroup.

I think it can only be guaranteed by using the [`WaveSize` attribute](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_WaveSize.html).

Reply:
Ah, ok. As far as I am aware, heterogeneous subgroup sizes happen when a compute shader dispatch knows the grid / workitem counts and then partially deactivates the subgroups around the edges of the grid. But DirectX only knows the workgroup count at compute shader dispatch, unlike e.g. OpenCL. So it shouldn't be a problem for the wgpu backend anyway. Additionally the [DirectX wiki](https://github.com/Microsoft/DirectXShaderCompiler/wiki/Wave-Intrinsics) says that `WaveGetLaneCount()` is unaffected by partial deactivation.

Reply:
My worry is that instead of launching same sized subgroups with inactive lanes, different sized subgroups would be launched (still not necessarily with all lanes active but at least less inactive ones).

The only indication of the former is the DXC wiki mentioning

> Returns the number of lanes in a wave on this architecture.

Reply:
https://github.com/microsoft/DirectXShaderCompiler/issues/4731 & https://github.com/microsoft/hlsl-specs/issues/16#issuecomment-1636547485 do make it sound like all subgroups should have the same size. But since "driver-defined constants" are not yet a thing and `WaveGetLaneCount()` is just a function it would be conceivable that some driver would launch different sized subgroups.

I'm not sure what the best solution would be here. Maybe we should at least have a test making sure `subgroup_size` returns the same value for a bunch of odd variations of workgroup and dispatch sizes.

Reply:
Looking around some more, everyone seems to rely on `WaveGetLaneCount()` returning a static value.

But Jesse Natalie says:

> > then you can get index of wave inside group by dividing group thread id by wave width
>
> No, there's no such guaranteed mapping

_from https://discord.com/channels/590611987420020747/996417435374714920/1084919934125088868_

Then says how VkOn12 works around this.

I looked at mesa's dozen driver (aka VkOn12) and the polyfill looks similar to the one they shared on discord with the only difference being that it also just divides the local invocation index by the subgroup size but only for 1D workgroup sizes.

Here is the permalink:
https://gitlab.freedesktop.org/mesa/mesa/-/blob/c2ec0111b21667d1db9d6ceae295b033f044ed87/src/microsoft/compiler/dxil_nir.c#L1937

Looking at the blame, it seems the fastpath [was added](https://gitlab.freedesktop.org/mesa/mesa/-/commit/4f56cede6deff561a8cfa3fe893cf085cb8b2f1c) because DOOM Eternal's shaders were relying on that property 😅.

Reply:
I added a test to make sure the `subgroup_size` stays consistent within a workgroup.

Reply:
> So they implement ticked number drawing via atomic operations on workgroup memory.
Meaning their assignment of workitem-index in a workgroup to subgroup-index in a workgroup is chaotic / random.

That explains the "corruption/flickering".

> I added a test to make sure the `subgroup_size` stays consistent within a workgroup.

Thanks, it looks good!

So given that games are relying on this behavior I guess our mapping is probably fine (at least for 1D workgroup sizes).

My last question would be: What about 2D/3D workgroup sizes? Do we need the more complex polyfill dozen is using?

Reply:
Multidimensional cases are certainly tricky because there are many possible serializations and we don't know what the driver and the firmware on the GPU is going to use. Also, our tests currently only have 1D workgroups.

Reply:
I think we should (at least for now) not allow using `SubgroupId` in entry-points with non-1D workgroup sizes. What do you think?